Project #2: Automation of Concrete Mix Design (NDOT)
Team: Fatima, Lucas, Macy
Date: 2/24/26
Client: Nebraska Department of Transportation

Project Context (Client Summary)
NDOT has requested a Python-based automation of its Excel “Mix Design” worksheet logic. The workflow must:
- Translate Excel logic into Python functions.
- Prompt the user for inputs sequentially (mirroring Excel order).
- Generate a client-readable weight chart for one cubic yard.
- Evaluate four realistic mix scenarios as basic verification/validation.

Deliverables Checklist:
- Excel logic replicated using major calculation functions
- Sequential user prompts implemented
- Final weight chart for 1 yd³ generated and formatted
- 4 mix scenarios documented + run through the model
- Scenario outputs compared (quick verification/validation narrative)

Part 1 - Setup and Constants
1.1 Imports and Global Settings
Units convention:
- Cement, water, aggregates: lb/yd³
- Specific gravity: dimensionless
- Air content: %
- w/c ratio: dimensionless

In [15]:
cubic_yard_ft3 = 27 #cubic feet in one cubic yard
unit_weight_water = 62.4 #lb/ft^3 unit weight of water

1.2 Assumptions and Constants
- Assumption 1: Calculations are based on a fixed volume of 1 cubic yard (27 ft³).
- Assumption 2: All specific gravity values are based on the unit weight of water at 62.4 lb/ft³.

1.3 Inputs Dictionary
"cement_A": "Weight of Cement (lb/yd³)",
"fly_ash_B": "Weight of Fly Ash (lb/yd³)",
"silica_fume_C": "Weight of Silica Fume (lb/yd³)",
"other_scm_D": "Weight of other Supplementary Cementitious Materials (lb/yd³)",
"wc_ratio_E": "Target Water–Cement Ratio (dimensionless)",
"air_content_F": "Target Air Content (%)",
"percent_fine_G": "Target percent of fine aggregate (%)",
"percent_course_H": "Target percent of coarse aggregate (%)",
"percent_other_I": "Target percent of other aggregate (%)",
"sg_cement_J": "Specific Gravity of Cement",
"sg_fly_ash_K": "Specific Gravity of Fly Ash",
"sg_silica_fume_L": "Specific Gravity of Silica Fume",
"sg_other_scm_M": "Specific Gravity of other SCMs",
"sg_fine_N": "Specific Gravity of Fine Aggregate",
"sg_coarse_O": "Specific Gravity of Coarse Aggregate",
"sg_other_P": "Specific Gravity of Other Aggregate"

Part 2 - Engineering Functions
Each function replicates a formula from the NDOT Mix Design Excel sheet.

2.1 Function Q: Water Weight (lb/yd³)
Inputs: cement_A, fly_ash_B, silica_C, other_scm_D, wc_ratio_E
Output: water_weight_Q (lb/yd³)
Note: This functions is a replica of the Excel Formula "Q = (A + B + C + D) * E" to determine total water weight based on the water-to-cementitious ratio

In [16]:
def calculate_water_weight_Q(cement_A, fly_ash_B, silica_C, other_scm_D, wc_ratio_E): 
    water_weight_Q = (cement_A + fly_ash_B + silica_C + other_scm_D) * wc_ratio_E
    return water_weight_Q

2.2 Functions R, S, T, U: Cementitious Volumes (ft³)
Inputs: Weight of material (lb), Specific Gravity of material
Output: Volume of material (ft³)
Note: These functions convert weights into absolute volumes using the specific gravity and the unit weight of water (62.4 lb/ft^3)

In [17]:
def calculate_volume_R(cement_A, sg_cement_J):
    return cement_A / (sg_cement_J * unit_weight_water)
def calculate_fly_ash_volume_S(fly_ash_B, sg_fly_ash_K):
    return fly_ash_B / (sg_fly_ash_K * unit_weight_water)
def calculate_silica_volume_T(silica_C, sg_silica_L):
    return silica_C / (sg_silica_L * unit_weight_water)
def calculate_scm_volume_U(other_scm_D, sg_other_scm_M):
    return other_scm_D / (sg_other_scm_M * unit_weight_water)

2.3 Functions V & W: Air and Water Volumes (ft^3)
Inputs: air_content_F (%) or water_weight_Q (lb)
Output: Volume (ft^3)
Note: These functions calculate the space occupied by air and water within the 27 ft^3 total volume.

In [18]:
def air_volume_V(air_content_F):
    return (air_content_F / 100) * cubic_yard_ft3
def water_volume_W(water_weight_Q):
    return water_weight_Q / unit_weight_water

2.4 Function X: Total Aggregate Volume (ft^3)
Inputs: R, S, T, U, V, W (ft^3)
Output: volume_X (ft^3)
Note: This function determines the volume leftover available for aggregates by subtracting all other material volumes from 27 ft³

In [19]:
def calculate_aggregate_volume_X(volume_R, volume_S, volume_T, volume_U, volume_V, volume_W):
    volume_X = (cubic_yard_ft3 - volume_R - volume_S - volume_T - volume_U - volume_V - volume_W)
    return volume_X

2.5 Functions Y, Z, AA: Aggregate Weights (lb/yd^3)
Inputs: Target percentage (%), Specific Gravity, Total Aggregate Volume (X)
Output: Material Weight (lb)
Note: Converts the volume for each aggregate type back into batch weights

In [20]:
def calculate_fine_aggregate_Y(percent_fine_G, sg_fine_N, volume_X):
    return unit_weight_water * (percent_fine_G / 100) * sg_fine_N * volume_X
def calculate_course_aggregate_Z(percent_course_H, sg_course_O, volume_X):
    return unit_weight_water * (percent_course_H / 100) * sg_course_O * volume_X
def calculate_other_aggregate_AA(percent_other_I, sg_other_P, volume_X):
    return unit_weight_water * (percent_other_I / 100) * sg_other_P * volume_X

Part 3 - Sequential User Inputs

In [30]:
# Project Info
project_no = int(input("Enter project number: "))
concrete_class = input("Enter class of concrete: ")

# Cementitious material inputs
cement_A = float(input("Enter cement weight A in lb per cubic yard: "))
fly_ash_B = float(input("Enter fly ash weight B in lb per cubic yard: "))
silica_fume_C = float(input("Enter silica fume weight C in lb per cubic yard: "))
other_scm_D = float(input("Enter other SCM D in lb per cubic yard: "))

# Design parameters
wc_ratio_E = float(input("Enter water cement ratio E: "))
air_content_F = float(input("Enter air content F as a percentage: "))

# Aggregate proportions
percent_fine_G = float(input("Enter percent of fine aggregate G: "))
percent_course_H = float(input("Enter percent of course aggregate H: "))
percent_other_I = float(input("Enter percent of other aggregate I: "))

# Specific gravities
sg_cement_J = float(input("Enter specific gravity of cement J: "))
sg_fly_ash_K = float(input("Enter specific gravity of fly ash K: "))
sg_silica_fume_L = float(input("Enter specific gravity of silica fume L: "))
sg_other_scm_M = float(input("Enter specific gravity of other SCM M: "))
sg_fine_N = float(input("Enter specific gravity of fine aggregate N: "))
sg_coarse_O = float(input("Enter specific gravity of coarse aggregate O: "))
sg_other_P = float(input("Enter specific gravity of other aggregate P: "))

Enter project number:  04
Enter class of concrete:  PR3
Enter cement weight A in lb per cubic yard:  750
Enter fly ash weight B in lb per cubic yard:  0
Enter silica fume weight C in lb per cubic yard:  0
Enter other SCM D in lb per cubic yard:  0
Enter water cement ratio E:  0.38
Enter air content F as a percentage:  6.0
Enter percent of fine aggregate G:  40
Enter percent of course aggregate H:  60
Enter percent of other aggregate I:  0
Enter specific gravity of cement J:  3.15
Enter specific gravity of fly ash K:  2.15
Enter specific gravity of silica fume L:  2.20
Enter specific gravity of other SCM M:  2.60
Enter specific gravity of fine aggregate N:  2.62
Enter specific gravity of coarse aggregate O:  2.68
Enter specific gravity of other aggregate P:  2.60


Part 4 - Running Complete Mix Design

In [31]:
# 1. Calculate Weights and Volumes
water_weight_Q = calculate_water_weight_Q(cement_A, fly_ash_B, silica_fume_C, other_scm_D, wc_ratio_E)
R = calculate_volume_R(cement_A, sg_cement_J)
S = calculate_fly_ash_volume_S(fly_ash_B, sg_fly_ash_K)
T = calculate_silica_volume_T(silica_fume_C, sg_silica_fume_L)
U = calculate_scm_volume_U(other_scm_D, sg_other_scm_M)
V = air_volume_V(air_content_F)
W = water_volume_W(water_weight_Q)
X = calculate_aggregate_volume_X(R, S, T, U, V, W)
Y = calculate_fine_aggregate_Y(percent_fine_G, sg_fine_N, X)
Z = calculate_course_aggregate_Z(percent_course_H, sg_coarse_O, X)
AA = calculate_other_aggregate_AA(percent_other_I, sg_other_P, X)

# 2. Output Weight Chart
print("\n---------------------------------------------")
print(" NDOT Concrete Mix Design – Weight Summary")
print("         (1 Cubic Yard of Concrete)")
print("---------------------------------------------")
print(f"Project Number:       {project_no}")
print(f"Class of Concrete:    {concrete_class}")
print("---------------------------------------------")
print(f"Cement (A):            {cement_A:8.1f} lb")
print(f"Fly Ash (B):           {fly_ash_B:8.1f} lb")
print(f"Silica Fume (C):       {silica_fume_C:8.1f} lb")
print(f"Other SCM (D):         {other_scm_D:8.1f} lb")
print("---------------------------------------------")
print(f"Fine Aggregate (Y):    {Y:8.1f} lb")
print(f"Coarse Aggregate (Z):  {Z:8.1f} lb")
print(f"Other Aggregate (AA):  {AA:8.1f} lb")
print("---------------------------------------------")
print(f"Water (Q):             {water_weight_Q:8.1f} lb")
print("---------------------------------------------")
print("End of Mix Design Summary")


---------------------------------------------
 NDOT Concrete Mix Design – Weight Summary
         (1 Cubic Yard of Concrete)
---------------------------------------------
Project Number:       4
Class of Concrete:    PR3
---------------------------------------------
Cement (A):               750.0 lb
Fly Ash (B):                0.0 lb
Silica Fume (C):            0.0 lb
Other SCM (D):              0.0 lb
---------------------------------------------
Fine Aggregate (Y):      1111.5 lb
Coarse Aggregate (Z):    1705.5 lb
Other Aggregate (AA):       0.0 lb
---------------------------------------------
Water (Q):                285.0 lb
---------------------------------------------
End of Mix Design Summary


Part 5 - Running Scenarios

Scenario 1: Standard Pavement (Class 47B)
Source: Nebraska Department of Transportation. (2017). Standard specifications for highway construction. Section 1002.
Case: Standard mainline highway paving.
Inputs:
Project Number: 01
Class of Concrete: 47B
Cement Weight A: 564
Fly Ash Weight B: 0
Silica Fume Weight C: 0
Other SCM D: 0
Water Cement Ratio E: 0.45
Air Content F: 6.5
Percent Fine G: 70
Percent Coarse H: 30
Percent Other I: 0
SG Cement J: 3.15
SG Fly Ash K: 2.15 (Placeholder, even if B is 0)
SG Silica L: 2.20
SG Other SCM M: 2.60
SG Fine N: 2.62
SG Coarse O: 2.66
SG Other P: 2.60
Scenario 01 Summary Table:
---------------------------------------------
 NDOT Concrete Mix Design – Weight Summary
         (1 Cubic Yard of Concrete)
---------------------------------------------
Project Number:       1
Class of Concrete:    47B
---------------------------------------------
Cement (A):               564.0 lb
Fly Ash (B):                0.0 lb
Silica Fume (C):            0.0 lb
Other SCM (D):              0.0 lb
---------------------------------------------
Fine Aggregate (Y):      2095.2 lb
Coarse Aggregate (Z):     911.7 lb
Other Aggregate (AA):       0.0 lb
---------------------------------------------
Water (Q):                253.8 lb
---------------------------------------------
End of Mix Design Summary

Scenario 2: Bridge Deck Optimized (Class 47BR)
Source: Morcous, G., & Hu, J. (2015). Optimized aggregate gradations for Portland cement concrete mix designs evaluation (Final Report No. P336). Nebraska Department of Transportation
Case: Bridge deck construction requiring ASR mitigation and high durability.
Inputs:
Project Number: 02
Class of Concrete: 47BR
Cement Weight A: 510
Fly Ash Weight B: 145
Silica Fume Weight C: 0
Other SCM D: 0
Water Cement Ratio E: 0.42
Air Content F: 7.5
Percent Fine G: 45
Percent Coarse H: 50
Percent Other I: 5
SG Cement J: 3.15
SG Fly Ash K: 2.38
SG Silica L: 2.20
SG Other SCM M: 2.60
SG Fine N: 2.64
SG Coarse O: 2.68
SG Other P: 2.65
Scenario 02 Summary Table:
---------------------------------------------
 NDOT Concrete Mix Design – Weight Summary
         (1 Cubic Yard of Concrete)
---------------------------------------------
Project Number:       2
Class of Concrete:    47BR
---------------------------------------------
Cement (A):               510.0 lb
Fly Ash (B):              145.0 lb
Silica Fume (C):            0.0 lb
Other SCM (D):              0.0 lb
---------------------------------------------
Fine Aggregate (Y):      1259.9 lb
Coarse Aggregate (Z):    1421.1 lb
Other Aggregate (AA):     140.5 lb
---------------------------------------------
Water (Q):                275.1 lb
---------------------------------------------
End of Mix Design Summary

Scenario 3: Low Carbon Concrete (Class LCCI)
Source: Morcous, G., & Hu, J. (2019). Evaluation of reducing cement content in NDOR class R combined aggregate gradations (Final Report No. M069). Nebraska Department of Transportation.
Case: Sustainable infrastructure trials with reduced cement content
Inputs:
Project Number: 03
Class of Concrete: LCCI
Cement Weight A: 470
Fly Ash Weight B: 120
Silica Fume Weight C: 25
Other SCM D: 0
Water Cement Ratio E: 0.40
Air Content F: 6.0
Percent Fine G: 50
Percent Coarse H: 50
Percent Other I: 0
SG Cement J: 3.15
SG Fly Ash K: 2.30
SG Silica L: 2.20
SG Other SCM M: 2.60
SG Fine N: 2.63
SG Coarse O: 2.65
SG Other P: 2.60
Scenario 03 Summary Table:
---------------------------------------------
 NDOT Concrete Mix Design – Weight Summary
         (1 Cubic Yard of Concrete)
---------------------------------------------
Project Number:       3
Class of Concrete:    LCCI
---------------------------------------------
Cement (A):               470.0 lb
Fly Ash (B):              120.0 lb
Silica Fume (C):           25.0 lb
Other SCM (D):              0.0 lb
---------------------------------------------
Fine Aggregate (Y):      1479.3 lb
Coarse Aggregate (Z):    1490.6 lb
Other Aggregate (AA):       0.0 lb
---------------------------------------------
Water (Q):                246.0 lb
---------------------------------------------
End of Mix Design Summary

Scenario 4: High Early Strength (Class PR3)
Source: Nebraska Department of Transportation. (2017). Standard specifications for highway construction. Section 1002.03
Case: Emergency pavement "patching" for rapid traffic reopening.
Inputs:
Project Number: 04
Class of Concrete: PR3
Cement Weight A: 750
Fly Ash Weight B: 0
Silica Fume Weight C: 0
Other SCM D: 0
Water Cement Ratio E: 0.38
Air Content F: 6.0
Percent Fine G: 40
Percent Coarse H: 60
Percent Other I: 0
SG Cement J: 3.15
SG Fly Ash K: 2.15
SG Silica L: 2.20
SG Other SCM M: 2.60
SG Fine N: 2.62
SG Coarse O: 2.68
SG Other P: 2.60
Scenario 04 Summary Table:
---------------------------------------------
 NDOT Concrete Mix Design – Weight Summary
         (1 Cubic Yard of Concrete)
---------------------------------------------
Project Number:       4
Class of Concrete:    PR3
---------------------------------------------
Cement (A):               750.0 lb
Fly Ash (B):                0.0 lb
Silica Fume (C):            0.0 lb
Other SCM (D):              0.0 lb
---------------------------------------------
Fine Aggregate (Y):      1111.5 lb
Coarse Aggregate (Z):    1705.5 lb
Other Aggregate (AA):       0.0 lb
---------------------------------------------
Water (Q):                285.0 lb
---------------------------------------------
End of Mix Design Summary